File: crates/verify/ess-conformance/src/ts/fixtures.ts
// Independently supplied values resolved before scenario activity.
import { array, closed, goMarshal, isObject, SelectionObservation, strictJSON } from './runtime.js';
import type { AccessorField, Node, SelectionDeclaration } from './runtime.js';
import { admitTypedDeclarations, decodeDeclaration, validateTypedFields } from './response.js';

export interface FixtureContract {
  fields: AccessorField[];
  declarations: Record<string, SelectionDeclaration>;
}

export function fixtureName(value: unknown): asserts value is string {
  if (
    typeof value !== 'string' ||
    value.length > 128 ||
    !/^[a-z][a-z0-9]*(?:-[a-z0-9]+)*$/.test(value)
  ) {
    throw new Error('invalid fixture name');
  }
}

export function admitFixtures(value: unknown): FixtureContract {
  const root = closed(value, 'fields declarations', '');
  admitTypedDeclarations(root.declarations);
  const fields = array(root.fields).map((raw) => {
    const field = closed(raw, 'name type', '');
    fixtureName(field.name);
    if (typeof field.type !== 'string') throw new Error('fixture type must be text');
    return { name: field.name, type: field.type };
  });
  const declarations = Object.fromEntries(
    Object.entries(root.declarations as Record<string, Node>).map(([key, body]) => [
      key,
      decodeDeclaration(body),
    ]),
  );
  if (fields.length === 0 || fields.length > 256 || Object.keys(declarations).length > 4096) {
    throw new Error('fixture field/declaration bound');
  }
  validateTypedFields([fields], declarations);
  if (new TextEncoder().encode(goMarshal(value)).length > 1048576)
    throw new Error('fixture contract byte limit');
  return { fields, declarations };
}

/** Copies all nested values, preserving exact numeric tokens and rejecting non-wire data. */
export function copyFixtureValue(value: Node): Node {
  const raw = goMarshal(value);
  if (new TextEncoder().encode(raw).length > 1048576) throw new Error('fixture value byte limit');
  return strictJSON(raw);
}

export function fixtureValues(contract: FixtureContract, provided: Node): Record<string, Node> {
  // Validate the original result before copying: undefined or functions must not disappear.
  if (!isObject(provided)) throw new Error('fixture values must be an object');
  const expected = contract.fields.map((field) => field.name).sort();
  if (JSON.stringify(Object.keys(provided).sort()) !== JSON.stringify(expected)) {
    throw new Error('fixture result must contain exactly the declared names');
  }
  const observer = SelectionObservation.forResponse(contract.declarations);
  const counter = { bytes: 0 };
  // A round trip normalizes ordinary JS numbers into the runtime's exact numeric tokens.
  const values = copyFixtureValue(provided) as Record<string, Node>;
  if (JSON.stringify(Object.keys(values).sort()) !== JSON.stringify(expected)) {
    throw new Error('fixture result lost a declared value');
  }
  for (const field of contract.fields) {
    observer.validateValue(
      field.type,
      values[field.name],
      Object.hasOwn(values, field.name),
      counter,
      0,
    );
  }
  if (counter.bytes > 1048576) throw new Error('fixture value byte limit');
  return values;
}

/** Inspect only typed scenario-value positions; literal objects cannot declare references. */
export function admitFixtureSteps(steps: Node[]): void {
  const declared = new Set<string>();
  const referenced = new Set<string>();
  for (const [index, step] of steps.entries()) {
    if (step.step === 'resolve_fixtures') {
      if (index !== 0) throw new Error('fixture resolution must be the first and only prelude');
      for (const field of admitFixtures(step.fixtures).fields) declared.add(field.name);
    }
    const groups: Record<string, Node>[] = [];
    switch (step.step) {
      case 'execute_command':
      case 'expect_invocation':
        groups.push(step.input ?? {});
        break;
      case 'snapshot_subject':
        groups.push(step.subject ?? {});
        break;
      case 'query_view':
      case 'eventually_view':
      case 'expect_halt':
      case 'eventually_halt':
        groups.push(step.params ?? {});
        break;
      case 'expect_event_values':
        if (Object.keys(step.payload).length === 0) throw new Error('empty event value assertion');
        groups.push(step.payload);
        break;
    }
    if (step.step === 'expect_view' || step.step === 'eventually_view')
      groups.push(step.expectation.fields ?? {});
    for (const group of groups)
      for (const value of Object.values(group)) {
        if (value.kind === 'fixture') referenced.add(value.fixture);
      }
  }
  if (declared.size !== referenced.size || [...referenced].some((key) => !declared.has(key))) {
    throw new Error('fixture declarations must exactly match the referenced names');
  }
}

File: crates/verify/ess-conformance/src/go/fixtures.go
// FixtureContract is source-owned type authority, resolved before scenario activity.
type FixtureContract struct {
	Fields       []accessorField                 `json:"fields"`
	Declarations map[string]selectionDeclaration `json:"declarations"`
}

// FixtureProvider supplies independently provisioned values. It owns provisioning cleanup.
type FixtureProvider interface {
	FixtureValues(ScenarioContext, FixtureContract) (map[string]Node, error)
}

func fixtureName(value any) error {
	label, ok := value.(string)
	if !ok || len(label) > 128 || !regexp.MustCompile(`^[a-z][a-z0-9]*(?:-[a-z0-9]+)*$`).MatchString(label) {
		return fmt.Errorf("invalid fixture name")
	}
	return nil
}

func (c *FixtureContract) UnmarshalJSON(raw []byte) error {
	type plain FixtureContract
	var value plain
	decoder := json.NewDecoder(strings.NewReader(string(raw)))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(&value); err != nil {
		return err
	}
	*c = FixtureContract(value)
	return c.validate()
}

func (c FixtureContract) validate() error {
	if len(c.Fields) == 0 || len(c.Fields) > 256 || len(c.Declarations) > 4096 {
		return fmt.Errorf("fixture field/declaration bound")
	}
	for _, field := range c.Fields {
		if err := fixtureName(field.Name); err != nil {
			return err
		}
	}
	if err := validateTypedFields([][]accessorField{c.Fields}, c.Declarations); err != nil {
		return err
	}
	raw, err := json.Marshal(c)
	if err != nil {
		return err
	}
	if len(raw) > 1048576 {
		return fmt.Errorf("fixture contract byte limit")
	}
	return nil
}

func admitFixtures(value any) error {
	root, err := closed(value, "fields declarations", "")
	if err != nil {
		return err
	}
	if err := admitTypedDeclarations(root["declarations"]); err != nil {
		return err
	}
	fields, err := array(root["fields"])
	if err != nil {
		return err
	}
	for _, raw := range fields {
		field, err := closed(raw, "name type", "")
		if err != nil {
			return err
		}
		if err := fixtureName(field["name"]); err != nil {
			return err
		}
		if _, err := text(field["type"]); err != nil {
			return err
		}
	}
	raw, err := json.Marshal(value)
	if err != nil {
		return err
	}
	var contract FixtureContract
	return json.Unmarshal(raw, &contract)
}

func copyFixtureValue(value Node) (Node, error) {
	raw, err := json.Marshal(value)
	if err != nil {
		return nil, err
	}
	if len(raw) > 1048576 {
		return nil, fmt.Errorf("fixture value byte limit")
	}
	decoder := json.NewDecoder(strings.NewReader(string(raw)))
	decoder.UseNumber()
	var copied Node
	err = decoder.Decode(&copied)
	return copied, err
}

func (c FixtureContract) values(provided map[string]Node) (map[string]Node, error) {
	if err := c.validate(); err != nil {
		return nil, err
	}
	if len(provided) != len(c.Fields) {
		return nil, fmt.Errorf("fixture result must contain exactly the declared names")
	}
	copied, err := copyFixtureValue(provided)
	if err != nil {
		return nil, err
	}
	values, ok := copied.(map[string]any)
	if !ok {
		return nil, fmt.Errorf("fixture values must be an object")
	}
	observer := selectionObservation{Declarations: c.Declarations, responseMode: true}
	bytes := 0
	for _, field := range c.Fields {
		value, exists := values[field.Name]
		if !exists {
			return nil, fmt.Errorf("missing fixture %s", field.Name)
		}
		if err := observer.validateValue(field.Type, value, exists, &bytes, 0); err != nil {
			return nil, err
		}
	}
	if bytes > 1048576 {
		return nil, fmt.Errorf("fixture value byte limit")
	}
	return values, nil
}

func (r *run) resolveFixtures(context ScenarioContext, contract *FixtureContract) error {
	provider, ok := r.target.(FixtureProvider)
	if !ok {
		return fmt.Errorf("%w: no pre-execution fixture provider", ErrUnsupported)
	}
	// Give the provider its own contract copy, never mutable assertion authority.
	raw, err := json.Marshal(contract)
	if err != nil {
		return err
	}
	var request FixtureContract
	if err := json.Unmarshal(raw, &request); err != nil {
		return err
	}
	provided, err := provider.FixtureValues(context, request)
	if err != nil {
		return err
	}
	r.fixtures, err = contract.values(provided)
	return err
}

func admitFixtureSteps(steps []any) error {
	declared, referenced := map[string]bool{}, map[string]bool{}
	for index, raw := range steps {
		step := raw.(map[string]any)
		if step["step"] == "resolve_fixtures" {
			if index != 0 {
				return fmt.Errorf("fixture resolution must be the first and only prelude")
			}
			contract := step["fixtures"].(map[string]any)
			for _, raw := range contract["fields"].([]any) {
				declared[raw.(map[string]any)["name"].(string)] = true
			}
		}
		groups := []any{}
		switch step["step"] {
		case "execute_command", "expect_invocation":
			groups = append(groups, step["input"])
		case "snapshot_subject":
			groups = append(groups, step["subject"])
		case "query_view", "eventually_view", "expect_halt", "eventually_halt":
			groups = append(groups, step["params"])
		case "expect_event_values":
			if len(step["payload"].(map[string]any)) == 0 {
				return fmt.Errorf("empty event value assertion")
			}
			groups = append(groups, step["payload"])
		}
		if step["step"] == "expect_view" || step["step"] == "eventually_view" {
			groups = append(groups, step["expectation"].(map[string]any)["fields"])
		}
		for _, raw := range groups {
			if group, ok := raw.(map[string]any); ok {
				for _, raw := range group {
					value := raw.(map[string]any)
					if value["kind"] == "fixture" {
						referenced[value["fixture"].(string)] = true
					}
				}
			}
		}
	}
	if len(declared) != len(referenced) {
		return fmt.Errorf("fixture declarations must exactly match the referenced names")
	}
	for key := range referenced {
		if !declared[key] {
			return fmt.Errorf("unbound fixture %s", key)
		}
	}
	return nil
}

func (r *run) expectEventValues(index int, step Step) bool {
	// Payload is decoded as Node for legacy literal steps; this step explicitly admits Value.
	raw, err := json.Marshal(step.Payload)
	if err != nil {
		return r.fail(index, "fixture event payload: %v", err)
	}
	decoder := json.NewDecoder(strings.NewReader(string(raw)))
	decoder.UseNumber()
	var values map[string]Value
	if err := decoder.Decode(&values); err != nil {
		return r.fail(index, "fixture event values: %v", err)
	}
	payload, ok := r.resolveAll(index, values)
	if !ok {
		return false
	}
	// Match the Rust runner: select the first direct occurrence by name, never by its values.
	for _, event := range r.last.DirectEvents {
		if event.Event != step.Event {
			continue
		}
		if !matches(event.Payload, payload) {
			return r.fail(index, "`%s` carried different fixture values", step.Event)
		}
		if reason := holds(event.Payload, step.Shape); reason != "" {
			return r.fail(index, "`%s` was emitted, and %s", step.Event, reason)
		}
		return true
	}
	return r.fail(index, "`%s` was not emitted", step.Event)
}
